lsm_tree.c

```diff
@@ -0,0 +1,333 @@
+			"holds %d segments.\n", lsm_tree->memtable->count_keys,
+			lsm_tree->full_segments);
+	print_active_segments(lsm_tree);
+}
+
+void print_active_segments(LSM_Tree *lsm_tree) {
+	for (int i = 0; i < lsm_tree->full_segments; i++) {
+		printf("%s\n", lsm_tree->segments[i]);
+	}
+}
+
+/* Call to deallocate all memory for LSM tree system*/
+void shutdown_lsm_system(LSM_Tree *lsm_tree) {
+	delete_memtable(lsm_tree->memtable);
+	free_segments(lsm_tree->segments);
+	free(lsm_tree);
+}
+
+static char* compact(char **segment_files) {
+	printf("> LSM System Alert: Running compaction...\n");
+
+	char *segment_a = *(segment_files);
+	char *new_segment, *segment_b;
+
+	for (int i = 1; i < MAX_SEGMENTS; i++) {
+		segment_b = *(segment_files + i);
+		new_segment = merge_files(segment_a, segment_b);
+
+		// old segment files no longer needed if new one was
+		// successfully created
+		if (delete_file(segment_a) != 0 || delete_file(segment_b) != 0)
+			printf("Failed to delete old segment files!\n");
+
+		// new segment will then be merged with consecutive other files
+		segment_a = new_segment;
+	}
+	return new_segment;
+}
+
+/* Used to perform compaction step of two segment files. This method is
+ * necessary for cleaning up old segment files and keeping read I/O from
+ * getting out of control. Merges old segments together into new segments.*/
+static char* merge_files(char *filename_a, char *filename_b) {
+	// open files for segments of interest
+	printf("Merging files %s and %s...\n", filename_a, filename_b);
+
+	FILE *seg_ptr_a = fopen(filename_a, "r");
+	FILE *seg_ptr_b = fopen(filename_b, "r");
+
+	// create a new segment file
+	char *new_segment = (char*) malloc(sizeof(char) * FILENAME_SIZE);
+	sprintf(new_segment, "%ld_%d.log", time(NULL), rand() % 10);
+	FILE *new_fp = fopen(new_segment, "w");
+
+	// setup for merge loop
+	char line_a[MAX_LINE_SIZE];
+	char line_b[MAX_LINE_SIZE];
+	bool keep_merging = true;
+	bool incr_ptr_a = true, incr_ptr_b = true;
+
+	// run the merge loop
+	while (keep_merging) {
+		if (incr_ptr_a) {
+			memset(line_a, '\0', MAX_LINE_SIZE);
+			readline_from_file(line_a, MAX_LINE_SIZE, seg_ptr_a, true);
+		}
+		if (incr_ptr_b) {
+			memset(line_b, '\0', MAX_LINE_SIZE);
+			readline_from_file(line_b, MAX_LINE_SIZE, seg_ptr_b, true);
+		}
+
+		if (!strlen(line_a) && !strlen(line_b))
+			keep_merging = false;
+
+		else if (!strlen(line_a) || !strlen(line_b)) {
+			// add any remaining keys in remaining file to new file
+			FILE *temp_ptr = strlen(line_a) ? seg_ptr_a : seg_ptr_b;
+			char *temp_line = strlen(line_a) ? line_a : line_b;
+			do {
+				fputs(temp_line, new_fp);
+				fputs("\n", new_fp);
+			} while (fgets(temp_line, MAX_LINE_SIZE, temp_ptr) != NULL);
+
+			keep_merging = false;
+
+		} else
+			merge_lines(line_a, line_b, new_fp, &incr_ptr_a, &incr_ptr_b);
+	}
+	// close file pointers for new file and the two segment files
+	fclose(new_fp);
```

Comment:
Let's handle the errors here? Same for other uses of ```fclose```

Reply:
ya, you're right I should check. I added error checking for the important cases.